File: consorcio/lib/domain/services/geo_distance_service.dart
import 'dart:math' as math;

double distanceMeters({
  required double latitudeA,
  required double longitudeA,
  required double latitudeB,
  required double longitudeB,
}) {
  const earthRadius = 6371000.0;
  double toRad(double value) => value * math.pi / 180;
  final dLat = toRad(latitudeB - latitudeA);
  final dLng = toRad(longitudeB - longitudeA);
  final lat1 = toRad(latitudeA);
  final lat2 = toRad(latitudeB);
  final haversine = math.sin(dLat / 2) * math.sin(dLat / 2) +
      math.cos(lat1) * math.cos(lat2) * math.sin(dLng / 2) * math.sin(dLng / 2);
  return 2 * earthRadius * math.asin(math.min(1, math.sqrt(haversine)));
}

class GeoBoundingBox {
  const GeoBoundingBox({
    required this.minLat,
    required this.maxLat,
    required this.minLng,
    required this.maxLng,
  });

  final double minLat;
  final double maxLat;
  final double minLng;
  final double maxLng;
}

GeoBoundingBox boundingBox({
  required double latitude,
  required double longitude,
  required double radiusMeters,
}) {
  const metersPerDegreeLat = 111320.0;
  final latDelta = radiusMeters / metersPerDegreeLat;
  final lngDelta = radiusMeters /
      (metersPerDegreeLat *
          math.max(0.2, math.cos(latitude * math.pi / 180)));
  return GeoBoundingBox(
    minLat: latitude - latDelta,
    maxLat: latitude + latDelta,
    minLng: longitude - lngDelta,
    maxLng: longitude + lngDelta,
  );
}

File: src/domain/services/GeoDistanceService.ts
/** Distancia en metros entre dos coordenadas (Haversine). */
export function distanceMeters(
  latitudeA: number,
  longitudeA: number,
  latitudeB: number,
  longitudeB: number,
): number {
  const earthRadius = 6_371_000
  const toRad = (value: number) => (value * Math.PI) / 180
  const dLat = toRad(latitudeB - latitudeA)
  const dLng = toRad(longitudeB - longitudeA)
  const lat1 = toRad(latitudeA)
  const lat2 = toRad(latitudeB)
  const haversine =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2
  return 2 * earthRadius * Math.asin(Math.min(1, Math.sqrt(haversine)))
}

const METERS_PER_DEGREE_LAT = 111_320

export interface GeoBoundingBox {
  minLat: number
  maxLat: number
  minLng: number
  maxLng: number
}

/** Caja alrededor de un punto, para consultar Firestore por lat/lng. */
export function boundingBox(
  latitude: number,
  longitude: number,
  radiusMeters: number,
): GeoBoundingBox {
  const latDelta = radiusMeters / METERS_PER_DEGREE_LAT
  const lngDelta =
    radiusMeters /
    (METERS_PER_DEGREE_LAT * Math.max(0.2, Math.cos((latitude * Math.PI) / 180)))
  return {
    minLat: latitude - latDelta,
    maxLat: latitude + latDelta,
    minLng: longitude - lngDelta,
    maxLng: longitude + lngDelta,
  }
}
